util/string_ref.c

```diff
@@ -0,0 +1,125 @@
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#include <plhash.h>
+#include <assert.h>
+#include <pthread.h>
+#include "string_ref.h"
+#include "locks_wrap.h"
+#include "comdb2_atomic.h"
+#include "logmsg.h"
+
+#define TRACK_REFERENCES
+
+#ifdef TRACK_REFERENCES
+static hash_t *hrefs = NULL;
+static pthread_mutex_t hmtx = PTHREAD_MUTEX_INITIALIZER;
+#endif
+int gbl_creation_count;
+
+struct string_ref_t {
+    char *str;
+    int cnt;
+};
+
+
+/* Makes a copy of the string passed and uses that as a reference counted object
+ */
+string_ref_t * create_string_ref(const char *str)
+{
+    assert(str);
+    int len = strlen(str);
+    string_ref_t *ref = malloc(sizeof(string_ref_t) + len + 1);
+    ref->cnt = 1;
+    ref->str = ((char*)ref) + sizeof(string_ref_t);
+    strcpy(ref->str, str);
+
+#ifdef TRACK_REFERENCES
+    Pthread_mutex_lock(&hmtx);
+    if(!hrefs) 
+        hrefs = hash_init(sizeof(&ref));
+    hash_add(hrefs, ref);
+    gbl_creation_count += 1;
+    Pthread_mutex_unlock(&hmtx);
+#else
+    ATOMIC_ADD32(gbl_creation_count, 1);
+#endif
+    return ref;
+}
+
+/* Get a reference by increasing the count */
+string_ref_t * get_ref(string_ref_t *ref)
+{
+    assert(ref);
+    int cnt = ATOMIC_ADD32(ref->cnt, 1);
+    if(cnt <= 1) // create has a reference, this can only be > 1
+        abort();
+    return ref;
+}
+
+/* Release a reference and free if this is the last holder */
+void put_ref(string_ref_t **ref)
+{
+    int cnt = ATOMIC_ADD32((*ref)->cnt, -1);
+    if(cnt < 0)
+        abort();
+    if (cnt == 0) {
+#ifdef TRACK_REFERENCES
+        Pthread_mutex_lock(&hmtx);
+        hash_del(hrefs, *ref);
+        gbl_creation_count -= 1;
+        Pthread_mutex_unlock(&hmtx);
+#else
+        ATOMIC_ADD32(gbl_creation_count, -1);
+#endif
+        free(*ref);
+    }
+    *ref = NULL;
+}
+
+const char *get_string(string_ref_t *ref)
+{
+    return ref->str;
+}
+
+#ifdef TRACK_REFERENCES
+static int print_it(void *obj, void *arg)
+{
+    string_ref_t *ref = obj;
+    logmsg(LOGMSG_USER, "%s:%d\n", ref->str, ref->cnt);
+    return 0;
+}
+
+static void print_all_string_references()
+{
+    if (gbl_creation_count > 0) {
+        Pthread_mutex_lock(&hmtx);
+        logmsg(LOGMSG_USER, "Remaining not-cleaned-up string references:\n");
+        hash_for(hrefs, print_it, NULL);
+        Pthread_mutex_unlock(&hmtx);
+    }
+}
+#endif
+
+int all_string_references_cleared()
+{
+    int res = (gbl_creation_count == 0);
```

Comment:
Should this line be moved to the #ifdef block below since you aren't modifying `gbl_creation_count` unless `TRACK_REFERENCES` is defined.

Reply:
It is modified--it's in the #else part of the get_ref() and put_ref(). The idea of TRACK_REFERENCES was to make sure the ref accounting is correct: that we release the references correctly and there are no leaks. Once we are certain, then we don't have to keep the hash tbl, but still abort on exit if the count is wrong so we can investigate.
For now, we want to print all remaining entries in the hash tbl on exit, and abort for visibility.